pkg/madmin/lock-commands.go

```diff
@@ -0,0 +1,136 @@
+	"time"
+)
+
+// LockDetail - represents lock specific details.
+type LockDetail struct {
+	OperationID string        // Unique ID per operation
+	Source      string        // Operation type (GetObject, PutObject...)
+	Type        string        // Lock type (RLock, WLock)
+	Status      string        // Status can be Running/Ready/Blocked
+	Since       time.Time     // Time when the lock was initially held
+	Duration    time.Duration // Duration the lock was held for
+}
+
+// LockInfo - represents summary and individual lock details of all
+// locks held on a given bucket, object.
+type LockInfo struct {
+	Bucket string
+	Object string
+	// Total no. of locks matching (incl. blocked) on bucket,
+	// object.
+	Total int64
+	// No. of locks acquired on bucket, object.
+	Acquired int64
+	// No. of locks blocked on bucket, object.
+	Blocked int64
+	// Details on API operation that attempted the lock, when, for
+	// how long etc.
+	Details []LockDetail
+}
+
+// getLockInfos - unmarshal []LockInfo from a reader.
+func getLockInfos(body io.Reader) ([]LockInfo, error) {
+	respBytes, err := ioutil.ReadAll(body)
+	if err != nil {
+		return nil, err
+	}
+
+	var lockInfos []LockInfo
+
+	err = json.Unmarshal(respBytes, &lockInfos)
+	if err != nil {
+		return nil, err
+	}
+
+	return lockInfos, nil
+}
+
+// ListLocks - Calls List Locks Management API to fetch locks matching
+// bucket, prefix and held before the duration supplied.
+func (adm *AdminClient) ListLocks(bucket, prefix string, olderThan time.Duration) ([]LockInfo, error) {
+	queryVal := make(url.Values)
+	queryVal.Set("lock", "")
+	queryVal.Set("bucket", bucket)
+	queryVal.Set("prefix", prefix)
+	queryVal.Set("older-than", olderThan.String())
+
+	hdrs := make(http.Header)
+	hdrs.Set(minioAdminOpHeader, "list")
+
+	reqData := requestData{
+		queryValues:   queryVal,
+		customHeaders: hdrs,
+	}
+
+	// Execute GET on /?lock to list locks.
+	resp, err := adm.executeMethod("GET", reqData)
+
+	defer closeResponse(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.New("Got " + resp.Status)
+	}
+
+	return getLockInfos(resp.Body)
+}
+
+// ClearLocks - Calls Clear Locks Management API to clear locks held
+// on bucket, matching prefix older than duration supplied.
+func (adm *AdminClient) ClearLocks(bucket, prefix string, olderThan time.Duration) ([]LockInfo, error) {
+	queryVal := make(url.Values)
+	queryVal.Set("lock", "")
+	queryVal.Set("bucket", bucket)
+	queryVal.Set("prefix", prefix)
+	queryVal.Set("older-than", olderThan.String())
+
+	hdrs := make(http.Header)
+	hdrs.Set(minioAdminOpHeader, "clear")
+
+	reqData := requestData{
+		queryValues:   queryVal,
+		customHeaders: hdrs,
+	}
+
+	// Execute POST on /?lock to clear locks.
+	resp, err := adm.executeMethod("POST", reqData)
+
+	defer closeResponse(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.New("Got " + resp.Status)
```

Comment:
OK, I shall change it uniformly in all places in pkg/madmin incl. service commands.